### PR TITLE
pre_defined_classes-simple_programs_taking_string_user_input-hard-q6

### PR DESCRIPTION
**Pull request by yuqi88**
## PR Checklist
- [ ]  Are helper methods really needed here?
    > Opt for hardcoding input/output into the input provider when possible
- [ ]  Are the assertion fail messages good?
    - [ ]  Logical (does it make sense?)
    - [ ]  Readable (can a student understand what’s wrong?)
    - [ ]  Not too feed-y (does it give away the answer?)
    - [ ]  Grammatically sound (is the grammar and punctuation in the sentence proper?)
    - [ ]  Specific (does the fail message relate to the question?)
- [ ]  Is the code formatted well?
    > Run `CTRL` + `ALT` + `L` in IntelliJ
- [ ]  EOF newlines
- [ ]  Do the comments for Parsons follow the correct convention?
- [ ]  Is the question being tested sufficiently?
    - [ ]  Are there representative cases to test the program?
    - [ ]  What about edge cases? (empty strings, negatives, infinity, off-by-one errors, etc)
        > If the question doesn’t specify what to do, mark it as a bad question
- [ ]  Are the test method names good?
    > Test method names show up in a “What went well” section on the website, so “mathTest” will become “What went well: math test”, which doesn’t make much sense. A better name in this example would be “calculatesTaxCorrectly” so it shows up as “What went well: calculates tax correctly”
---
## Website
- [ ] Is the question text on the Gamification document up to date? 
    > Make sure to include the sample output if there is one for the question

### CODE REVIEW

File: src/pre_defined_classes/simple_programs_taking_string_user_input/hard/q6/CanValidator.java [hunk 1]
```diff
@@ -0,0 +1,37 @@
+package pre_defined_classes.simple_programs_taking_string_user_input.hard.q6;
+
+/*
+Emailing is the most popular asynchronous communication tool in the 20th century so far. Suppose all UBC emails must
```
Comment (Hedgemon4) [inline src/pre_defined_classes/simple_programs_taking_string_user_input/hard/q6/CanValidator.java:4]: 21st century

File: src/pre_defined_classes/simple_programs_taking_string_user_input/hard/q6/MainTest.java [hunk 2]
```diff
@@ -0,0 +1,59 @@
...
+        return Stream.of("example2@gmail.com", "example3@ubc.com", "example3@gamil.ca",
+                "notValidEmailFormat", System.lineSeparator()
+        );
+    }
+
+    @ParameterizedTest
+    @MethodSource("UBCEmail")
+    public void checksUBCEmail(String input) throws InvalidClauseException {
+        TestOption.incorrectStructureErrorMessage =
+                "Your program does not print the output correctly when the input is an UBC email.";
```
Comment (Hedgemon4) [inline src/pre_defined_classes/simple_programs_taking_string_user_input/hard/q6/MainTest.java:44]: `... print the correct output when ...`

File: src/pre_defined_classes/simple_programs_taking_string_user_input/hard/q6/MainTest.java [hunk 3]
```diff
@@ -0,0 +1,59 @@
...
+
+    public static Stream<String> nonUBCEmail() {
+        return Stream.of("example2@gmail.com", "example3@ubc.com", "example3@gamil.ca",
+                "notValidEmailFormat", System.lineSeparator()
+        );
+    }
+
+    @ParameterizedTest
+    @MethodSource("UBCEmail")
+    public void checksUBCEmail(String input) throws InvalidClauseException {
```
Comment (Hedgemon4) [inline src/pre_defined_classes/simple_programs_taking_string_user_input/hard/q6/MainTest.java:42]: `worksWithUBCEmail`

Comment (Hedgemon4) [reply src/pre_defined_classes/simple_programs_taking_string_user_input/hard/q6/MainTest.java:42]: Or `identifiesUBCEmailCorrectly`

File: src/pre_defined_classes/simple_programs_taking_string_user_input/hard/q6/MainTest.java [hunk 4]
```diff
@@ -0,0 +1,59 @@
...
+        TestOption.incorrectStructureErrorMessage =
+                "Your program does not print the output correctly when the input is an UBC email.";
+        runWithInput(input, new Clause[]{
+                new StringLiteral("true")
+        });
+    }
+
+    @ParameterizedTest
+    @MethodSource("nonUBCEmail")
+    public void checksNonUBCEmail(String input) throws InvalidClauseException {
```
Comment (Hedgemon4) [inline src/pre_defined_classes/simple_programs_taking_string_user_input/hard/q6/MainTest.java:52]: See other comments

File: src/pre_defined_classes/simple_programs_taking_string_user_input/hard/q6/MainTest.java [hunk 5]
```diff
@@ -0,0 +1,59 @@
...
+        TestOption.incorrectStructureErrorMessage =
+                "Your program does not print the output correctly when the input is not an UBC email.";
```
Comment (Hedgemon4) [inline src/pre_defined_classes/simple_programs_taking_string_user_input/hard/q6/MainTest.java:54]: See other method comments

**Review by Hedgemon4 — CHANGES_REQUESTED** [review]

File: src/pre_defined_classes/simple_programs_taking_string_user_input/hard/q6/MainTest.java [hunk 6]
```diff
@@ -0,0 +1,59 @@
...
+        return Stream.of("example2@gmail.com", "example3@ubc.com", "example3@gamil.ca",
+                "notValidEmailFormat", System.lineSeparator()
+        );
+    }
+
+    @ParameterizedTest
+    @MethodSource("UBCEmail")
+    public void identifiesUBCEmailCorrectly(String input) throws InvalidClauseException {
+        TestOption.incorrectStructureErrorMessage =
+                "Your program does not print the correct output when the input is an UBC email.";
```
Comment (Hedgemon4) [inline src/pre_defined_classes/simple_programs_taking_string_user_input/hard/q6/MainTest.java:44]:
```suggestion
                "Your program does not print the correct output when the input is a UBC email.";
```

File: src/pre_defined_classes/simple_programs_taking_string_user_input/hard/q6/MainTest.java [hunk 7]
```diff
@@ -0,0 +1,59 @@
...
+        runWithInput(input, new Clause[]{
+                new StringLiteral("true")
+        });
+    }
+
+    @ParameterizedTest
+    @MethodSource("nonUBCEmail")
+    public void identifiesNonUBCEmailCorrectly(String input) throws InvalidClauseException {
+        TestOption.incorrectStructureErrorMessage =
+                "Your program does not print the output correctly when the input is not an UBC email.";
```
Comment (Hedgemon4) [inline src/pre_defined_classes/simple_programs_taking_string_user_input/hard/q6/MainTest.java:54]:
```suggestion
                "Your program does not print the output correctly when the input is not a UBC email.";
```

**Review by Hedgemon4 — CHANGES_REQUESTED** [review]

File: src/pre_defined_classes/simple_programs_taking_string_user_input/hard/q6/CanValidator.java [hunk 8]
```diff
@@ -0,0 +1,37 @@
+package pre_defined_classes.simple_programs_taking_string_user_input.hard.q6;
+
+/*
+Emailing is the most popular asynchronous communication tool in the 21th century so far. Suppose all UBC emails must
```
Comment (Hedgemon4) [inline src/pre_defined_classes/simple_programs_taking_string_user_input/hard/q6/CanValidator.java:4]:
```suggestion
Emailing is the most popular asynchronous communication tool in the 21st century so far. Suppose all UBC emails must
```

**Review by Hedgemon4 — CHANGES_REQUESTED** [review]

**Review by Hedgemon4 — APPROVED** [review]
